cpp/src/experimental/graph.cu

```diff
@@ -0,0 +1,543 @@
+                       // memory if minor can be larger than 32 bit but minor - minor_first fits
+                       // within 32 bit
+                       p_indices[start + idx] =
+                         minor;  // overwrite the counter only if idx == degree - 1 (no race)
+                       p_weights[start + idx] = w;
+                     });
+  } else {
+    auto edge_first = thrust::make_zip_iterator(
+      thrust::make_tuple(edgelist.p_src_vertices, edgelist.p_dst_vertices));
+    thrust::for_each(rmm::exec_policy(handle.get_stream())->on(handle.get_stream()),
+                     edge_first,
+                     edge_first + edgelist.number_of_edges,
+                     [p_offsets, p_indices, p_weights, major_first] __device__(auto e) {
+                       auto s      = thrust::get<0>(e);
+                       auto d      = thrust::get<1>(e);
+                       auto major  = store_transposed ? d : s;
+                       auto minor  = store_transposed ? s : d;
+                       auto start  = p_offsets[major - major_first];
+                       auto degree = p_offsets[(major - major_first) + 1] - start;
+                       auto idx    = atomicAdd(p_indices + (start + degree - 1),
+                                            vertex_t{1});  // use the last element as a counter
+                       // FIXME: we can actually store minor - minor_first instead of minor to save
+                       // memory if minor can be larger than 32 bit but minor - minor_first fits
+                       // within 32 bit
+                       p_indices[start + idx] =
+                         minor;  // overwrite the counter only if idx == degree - 1 (no race)
+                     });
+  }
+
+  // FIXME: need to add an option to sort neighbor lists
+
+  return std::make_tuple(std::move(offsets), std::move(indices), std::move(weights));
+}
+
+template <typename vertex_t, typename DegreeIterator, typename ThresholdIterator>
+std::vector<vertex_t> segment_degree_sorted_vertex_partition(raft::handle_t const &handle,
+                                                             DegreeIterator degree_first,
+                                                             DegreeIterator degree_last,
+                                                             ThresholdIterator threshold_first,
+                                                             ThresholdIterator threshold_last)
+{
+  auto num_elements = thrust::distance(degree_first, degree_last);
+  auto num_segments = thrust::distance(threshold_first, threshold_last) + 1;
+
+  std::vector<vertex_t> h_segment_offsets(num_segments + 1);
+  h_segment_offsets[0]     = 0;
+  h_segment_offsets.back() = num_elements;
+
+  rmm::device_uvector<vertex_t> d_segment_offsets(num_segments - 1, handle.get_stream());
+
+  thrust::upper_bound(rmm::exec_policy(handle.get_stream())->on(handle.get_stream()),
```

Comment:
CAVEAT: thrust `lower_bound()` / `upper_bound()` / `binary_search()`: https://github.com/thrust/thrust/pull/1104

Reply:
Thanks but it seems like this does not apply to this case.